### PR TITLE
added sc-alt account control

### PR DESCRIPTION
**Pull request by pranchalsomani**

### CODE REVIEW

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 1]
```diff
@@ -0,0 +1,915 @@
+<###
+# Overview:
+    This script is used to create PIM role assignment(s) with SC-ALT account for critical permanent role assignment(s) and PIM non SCALT role assignment(s) remove critical permanent role assignment(s) for which the PIM role assignment(s) is successfully created in the Subscription or in resource group.
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:3]: To be modified to - "PIM non SCALT role assignment(s) **and** remove"

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 2]
```diff
@@ -0,0 +1,915 @@
...
+
+# Control ID:
+    Azure_Subscription_AuthZ_Dont_Grant_Persistent_Access , Azure_Subscription_AuthZ_Dont_Grant_Persistent_Access_RG
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:6]: Should this also include the SC-ALT Control ID?

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 3]
```diff
@@ -0,0 +1,915 @@
...
+
+# Display Name:
+Do not grant permanent access for critical subscription level roles , Do not grant permanent access for critical resource group level roles
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:9]: Please punctuate this - "PIM non SCALT role assignment(s) **and** remove"

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 4]
```diff
@@ -0,0 +1,915 @@
...
+
+# Prerequisites:
+    Owner and higher privileged role assignment on the Subscription is required and atleast one service adminstrator role assignment must be present on the subscription level.
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:12]: To be modified to - "PIM non SCALT role assignment(s) **and** remove"

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 5]
```diff
@@ -0,0 +1,915 @@
...
+
+# Steps performed by the script:
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:14]: We'll have to explicitly convey that rollback is not supported.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 6]
```diff
@@ -0,0 +1,915 @@
...
+
+# Examples:
+    To remediate:
+        1. To review the critical Permanent role assignment(s) on a Subscription that will be remediated:
+    
+           Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -DryRun
+
+        2. To review the critical Permanent role assignment(s) on the resource group that will be remediated:
+    
+           Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck  -resourcegroup -DryRun
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:38]: Please fix `-resourcegroup` as per the conventions followed with others.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 7]
```diff
@@ -0,0 +1,915 @@
...
+    To remediate:
+        1. Download the script.
+        2. Load the script in a PowerShell session. Refer https://aka.ms/AzTS-docs/RemediationscriptExcSteps to know more about loading the script.
+        3. Execute the script to create PIM role assignment(s) with SC-ALT account for critical permanent role assignment(s) and remove critical permanent role assignment(s) for which the PIM role assignment(s) is successfully created in the Subscription or at resource group level. Refer `Examples`, below.
+
+# Examples:
+    To remediate:
+        1. To review the critical Permanent role assignment(s) on a Subscription that will be remediated:
+    
+           Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -DryRun
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:34]: Just a thought - Can this be renamed to `Migrate-PermanentToPrivilegedRoleAssignments` or `Convert-PermanentToPrivilegedRoleAssignments`.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 8]
```diff
@@ -0,0 +1,915 @@
...
+
+        2. To review the critical Permanent role assignment(s) on the resource group that will be remediated:
+    
+           Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck  -resourcegroup -DryRun
+
+        3. To create PIM role assignment(s) with SC-ALT account for critical permanent role assignment(s) and PIM non SCALT role assignment(s) remove critical permanent role assignment(s) for which the PIM role assignment(s) is successfully created in the Subscription or in resource group, from a previously taken snapshot:
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:40]: To be modified to - "PIM non SCALT role assignment(s) **and** remove"

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 9]
```diff
@@ -0,0 +1,915 @@
...
+          
+           Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -FilePath C:\AzTS\Subscriptions\00000000-xxxx-0000-xxxx-000000000000\202109131040\        $backupFolderPath = "$([Environment]::GetFolderPath('LocalApplicationData'))\AzTS\Remediation\Subscriptions\$($context.Subscription.SubscriptionId.replace('-','_'))\$($(Get-Date).ToString('yyyyMMddhhmm'))\RemoveCriticalPermanentAndPIMRoleAssignments"\CriticalPermanentAndPIMRoleAssignment.csv
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:42]: `FilePath` to be fixed.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 10]
```diff
@@ -0,0 +1,915 @@
...
+
+        4. To create PIM role assignment(s) with SC-ALT account for critical permanent role assignment(s) and PIM non SCALT role assignment(s) remove critical permanent role assignment(s) for which the PIM role assignment(s) is successfully created in the Subscription or in resource group, from a previously taken snapshot:
+          
+           Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -resourcegroup -FilePath C:\AzTS\Subscriptions\00000000-xxxx-0000-xxxx-000000000000\202109131040\        $backupFolderPath = "$([Environment]::GetFolderPath('LocalApplicationData'))\AzTS\Remediation\Subscriptions\$($context.Subscription.SubscriptionId.replace('-','_'))\$($(Get-Date).ToString('yyyyMMddhhmm'))\RemoveCriticalPermanentAndPIMRoleAssignments"\CriticalPermanentAndPIMRoleAssignment.csv
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:46]: To be modified to - "PIM non SCALT role assignment(s) **and** remove"
`FilePath` to be fixed.
Please remove the extra newline.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 11]
```diff
@@ -0,0 +1,915 @@
...
+    # List of required modules
+    $requiredModules = @("Az.Accounts", "Az.Resources", "AzureAD")
+
+    Write-Host "Required modules: $($requiredModules -join ', ')" -ForegroundColor $([Constants]::MessageType.Info)
+    Write-Host "Checking if the required modules are present..."
+    # List of available modules
+    $availableModules = $(Get-Module -ListAvailable $requiredModules -ErrorAction Stop)
+
+    # Check if the required modules are installed.
+    if ($availableModules.Name -contains "Az.Resources")
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:88]: Instead of specially handling this, can we modify L80 to include minimum required versions for each of the modules and use them here.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 12]
```diff
@@ -0,0 +1,915 @@
...
+        }
+    }
+    else
+    {
+        Write-Host "Installing $($_) module...." -ForegroundColor $([Constants]::MessageType.Update)
+        Install-Module -Name "Az.Resources" -MinimumVersion 5.0.0 -RequiredVersion 5.0.0 -Scope CurrentUser -Repository 'PSGallery' -Force -ErrorAction Stop
+    }
+
+
+    if($availableModules.Name -contains "AZ.Accounts")
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:108]: Instead of specially handling this, can we modify L80 to include minimum required versions for each of the modules and use them here.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 13]
```diff
@@ -0,0 +1,915 @@
...
+             Install-Module -Name "Az.Accounts" -MinimumVersion 2.5.4 -RequiredVersion 2.5.4 -Scope CurrentUser -Repository 'PSGallery' -Force -ErrorAction Stop        
+        }    
+    }
+    else
+    {
+        Write-Host "Installing Az.Accounts module...." -ForegroundColor $([Constants]::MessageType.Update)
+        Install-Module -Name "Az.Accounts" -MinimumVersion 2.5.4 -RequiredVersion 2.5.4 -Scope CurrentUser -Repository 'PSGallery'  -Force -ErrorAction Stop
+    }
+
+    if($availableModules.Name -contains "AzureAD")
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:127]: Instead of specially handling this, can we modify L80 to include minimum required versions for each of the modules and use them here.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 14]
```diff
@@ -0,0 +1,915 @@
...
+        Install-Module -Name AzureAD -Scope CurrentUser -Repository "PSGallery" -Force -ErrorAction Stop
+    }
+}
+
+
+function Create-PIMForPermanentRoleAssignments
+{
+    <#
+        .SYNOPSIS
+        Remediates 'Azure_Subscription_AuthZ_Dont_Grant_Persistent_Access , Azure_Subscription_AuthZ_Dont_Grant_Persistent_Access_RG' Control.
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:143]: Should this also include the SC-ALT Control ID?

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 15]
```diff
@@ -0,0 +1,915 @@
...
+
+        .DESCRIPTION
+        Remediates 'Azure_Subscription_AuthZ_Dont_Grant_Persistent_Access , Azure_Subscription_AuthZ_Dont_Grant_Persistent_Access_RG' Control.
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:146]: Should this also include the SC-ALT Control ID?
To be modified to - "PIM non SCALT role assignment(s) **and** remove"
Please fix `-resourcegroup` as per the conventions followed with others.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 16]
```diff
@@ -0,0 +1,915 @@
...
+        Specifies a dry run of the actual remediation.
+        
+        .PARAMETER FilePath
+        Specifies the path to the file to be used as input for the remediation.
+
+        .INPUTS
+        None. You cannot pipe objects to Create-PIMForPermanentRoleAssignments.
+
+        .OUTPUTS
+        None. Create-PIMForPermanentRoleAssignments does not return anything that can be piped and used as an input to another command.
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:168]: Are we intentionally not publishing the `resourcegroup` option?

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 17]
```diff
@@ -0,0 +1,915 @@
...
+
+        .EXAMPLE
+        PS> Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -DryRun
+
+        .EXAMPLE
+        PS> Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -resourcegroup -DryRun
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:174]: Please fix `-resourcegroup` as per the conventions followed with others. Also, why is this required?

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 18]
```diff
@@ -0,0 +1,915 @@
...
+
+        .EXAMPLE
+        PS> Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -FilePath C:\AzTS\Subscriptions\00000000-xxxx-0000-xxxx-000000000000\202109131040\        $backupFolderPath = "$([Environment]::GetFolderPath('LocalApplicationData'))\AzTS\Remediation\Subscriptions\$($context.Subscription.SubscriptionId.replace('-','_'))\$($(Get-Date).ToString('yyyyMMddhhmm'))\RemoveCriticalPermanentAndPIMRoleAssignments"\CriticalPermanentAndPIMRoleAssignment.csv
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:177]: `FilePath` to be fixed.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 19]
```diff
@@ -0,0 +1,915 @@
...
+
+        .EXAMPLE
+        PS> Create-PIMForPermanentRoleAssignments -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -resourcegroup -FilePath C:\AzTS\Subscriptions\00000000-xxxx-0000-xxxx-000000000000\202109131040\        $backupFolderPath = "$([Environment]::GetFolderPath('LocalApplicationData'))\AzTS\Remediation\Subscriptions\$($context.Subscription.SubscriptionId.replace('-','_'))\$($(Get-Date).ToString('yyyyMMddhhmm'))\RemoveCriticalPermanentAndPIMRoleAssignments"\CriticalPermanentAndPIMRoleAssignment.csv
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:180]: `FilePath` to be fixed.

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]
Review is still in progress. These are comments for the interim.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 20]
```diff
@@ -0,0 +1,915 @@
...
+    #List of role assignment details 
+    $roleAssignmentDetails = @()
+
+    Write-Host "checking if subscription contains one or more service adminstrator role assignment(s)"...
+
+    $roleAssignmentDetails = Get-AzRoleAssignment -scope "/subscriptions/$($SubscriptionId)"
+    $serviceAdminstrator = $roleAssignmentDetails | Where-Object {$_.Scope -eq "/subscriptions/$($SubscriptionId)" -and ($_.RoleDefinitionName -split";" -contains "ServiceAdministrator" -or $_.RoleDefinitionName -split";" -contains "CoAdministrator")}
+    
+    # Checking the count of service adminstrator role assignment in the subscription.
+    if(($serviceAdminstrator |Measure-Object).count -lt 1)
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:279]: Please replace with `eq 0`.

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 21]
```diff
@@ -0,0 +1,915 @@
...
+    {
+        Write-Host "WARNING: This script can only be run when the subscription contains one or more service admin role assignment(s)." -ForegroundColor $([Constants]::MessageType.Warning)
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:281]: Please rephrase this as - "WARNING: This script can only be run when there is a Service Administrator configured for the Subscription."

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 22]
```diff
@@ -0,0 +1,915 @@
...
+        #return
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:282]: Please remove lines 282 to 287.

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 23]
```diff
@@ -0,0 +1,915 @@
...
+            {
+                $scALTMapping.Add($_.Id , @($true,$_.userPrincipalName))
+            }
+            else
+            {
+                $scALTMapping.Add($_.Id , @($false,""))
+            }
+        }
+
+##################################################        #this to
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-PTAControls.ps1:700]: Please remove this and the stray newlines.

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 (hunk 7 above)
Comment (pranchalsomani) [reply Scripts/RemediationScripts/Remediate-PTAControls.ps1:44]: Migrate-PermanentToPrivilegedRoleAssignments

File: Scripts/RemediationScripts/Remediate-PTAControls.ps1 [hunk 24]
```diff
@@ -0,0 +1,915 @@
...
+    if ([String]::IsNullOrWhiteSpace($FilePath))
+    {
+        Write-Host "Fetching all role assignment(s) in Subscription: $($context.Subscription.SubscriptionId)" -ForegroundColor $([Constants]::MessageType.Info)
+
+        # Get all Permanent assignment(s) in a Subscription
+        $scope = "/subscriptions/$($subscriptionId)"
+        $roleAssignmentDetails = Get-AzRoleAssignmentScheduleInstance -scope $scope -ErrorAction Stop
+
+        # Get all PIM eligible assignment(s) in a subscription
+        $pimEligibleRoleAssignments = Get-AzRoleEligibilityScheduleInstance -Scope $scope
```
[the remark this answers is not in the document]
Comment (pranchalsomani) [reply Scripts/RemediationScripts/Remediate-PTAControls.ps1]: yes

**Review by Aboli-msft — APPROVED** [review]
